ta/sdp_basic/ta_sdp_basic.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2015, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#define CFG_CACHE_API
+
+#include <string.h>
+#include <tee_api.h>
+#include <tee_internal_api_extensions.h>
+#include <tee_internal_api.h>
+#include <tee_ta_api.h>
+#include <trace.h>
+
+#include <ta_sdp_basic.h>
+
+/*
+ * Basic Secure Data Path access test commands:
+ * - command INJECT: copy from non secure input into secure output.
+ * - command TRANSFROM: read secure input, transform and write tosecure output.
+ * - command DUMP: copy from secure input into non secure output.
+ */
+
+static TEE_Result cmd_inject(uint32_t types, TEE_Param params[4])
+{
+	TEE_Result rc;
+	int sec_idx;
+	int ns_idx;
+
+	/*
+	 * param#0: can be input our output buffer
+	 * param#1: can be input our output buffer
+	 */
+	switch (types) {
+	case TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+			     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+			     TEE_PARAM_TYPE_NONE, TEE_PARAM_TYPE_NONE):
+		ns_idx = 0;
+		sec_idx = 1;
+		break;
+	case TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+			     TEE_PARAM_TYPE_MEMREF_INPUT,
+			     TEE_PARAM_TYPE_NONE, TEE_PARAM_TYPE_NONE):
+		sec_idx = 0;
+		ns_idx = 1;
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (params[sec_idx].memref.size < params[ns_idx].memref.size)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	/*
+	 * check input/output buffers secure attribute:
+	 * Trust core on validity of the memref size: test only 1st byte
```

Comment:
Only first byte? We would like to check the entire range, don't we? Looks like you also are using ` params[sec|ns_idx].memref.size` here and not `1`.